### PR TITLE
Support shutdown write and notify for success write

### PR DESCRIPTION
**Pull request by chenBright**
### What problem does this PR solve?

Issue Number:

Problem Summary:

### What is changed and the side effects?

Changed:

1. 支持半关闭：无需SetFailed，关闭写入端，但不影响读数据。
2. 支持在数据写入socket写入成功后调用id_wait回调函数：对于一些请求和响应不是一一对应的场景，例如WebSocket、TCP Stream客户端、Streaming rpc等，只需要关注数据是否写入成功。数据写入socket写入成功后调用id_wait回调函数，可以统计成功率、耗时等。#2545 

Side effects:
- Performance effects(性能影响):

- Breaking backward compatibility(向后兼容性): 

---
### Check List:
- Please make sure your changes are compilable(请确保你的更改可以通过编译).
- When providing us with a new feature, it is best to add related tests(如果你向我们增加一个新的功能, 请添加相关测试).
- Please follow [Contributor Covenant Code of Conduct](../../master/CODE_OF_CONDUCT.md).(请遵循贡献者准则).

### CODE REVIEW

**Comment by chenBright** [comment]
@wwbmmm 有空看看这个PR

File: src/brpc/socket.cpp [hunk 1]
```diff
@@ -307,33 +307,56 @@ const uint32_t MAX_PIPELINED_COUNT = 16384;
...
-        return (_pc_and_udmsg >> 48) & 0x3FFF;
+        return _pc_and_udmsg.extra() & 0x3FFF;
     }
     uint32_t get_auth_flags() const {
-       return (_pc_and_udmsg >> 62) & 0x03;
+       return (_pc_and_udmsg.extra() >> 14) & 0x03;
     }
     void clear_pipelined_count_and_auth_flags() {
-        _pc_and_udmsg &= 0xFFFFFFFFFFFFULL;
+        _pc_and_udmsg.reset_ptr_and_extra();
```
Comment (wwbmmm) [inline src/brpc/socket.cpp:346]: 是不是应该是reset_extra() ？

File: src/brpc/socket.h [hunk 2]
```diff
@@ -917,6 +994,8 @@ friend void DereferenceSocket(Socket*);
     // Storing data that are not flushed into `fd' yet.
     butil::atomic<WriteRequest*> _write_head;
 
+    bool _is_wirte_shutdown;
```
Comment (wwbmmm) [inline src/brpc/socket.h:997]: wirte -> write

File: src/brpc/socket.cpp (hunk 1 above)
Comment (chenBright) [reply src/brpc/socket.cpp]: done

File: src/brpc/socket.h (hunk 2 above)
Comment (chenBright) [reply src/brpc/socket.h]: done

**Comment by wwbmmm** [comment]
LGTM

**Pull request merged**